src/settings/SettingsStore.ts

```diff
@@ -718,25 +718,37 @@ export default class SettingsStore {
 
     /**
      * Migrate the setting for visible images to a setting.
+     *
+     * @param isFreshLogin True if the user has just logged in, false if a previous session is being restored.
      */
-    private static migrateMediaControlsToSetting(): void {
-        const MIGRATION_DONE_FLAG = "mx_migrate_media_controls";
-        if (localStorage.getItem(MIGRATION_DONE_FLAG)) return;
-
-        logger.info("Performing one-time settings migration of show images and invite avatars to account data");
-        const handler = LEVEL_HANDLERS[SettingLevel.ACCOUNT];
-        const showImages = handler.getValue("showImages", null);
-        const showAvatarsOnInvites = handler.getValue("showAvatarsOnInvites", null);
+    private static migrateMediaControlsToSetting(isFreshLogin: boolean): void {
+        if (isFreshLogin) return;
+        const client = MatrixClientPeg.safeGet();
 
-        const AccountHandler = LEVEL_HANDLERS[SettingLevel.ACCOUNT];
-        if (showImages !== null || showAvatarsOnInvites !== null) {
-            AccountHandler.setValue("mediaPreviewConfig", null, {
-                invite_avatars: showAvatarsOnInvites === false ? MediaPreviewValue.Off : MediaPreviewValue.On,
-                media_previews: showImages === false ? MediaPreviewValue.Off : MediaPreviewValue.On,
-            });
-        } // else, we don't set anything and use the server value
+        const doMigration = async (): Promise<void> => {
+            while (!client.isInitialSyncComplete()) {
+                await new Promise((r) => client.once(ClientEvent.Sync, r));
+            }
+            // Never migrate if the config already exists.
+            if (client.getAccountData("io.element.msc4278.media_preview_config")) {
+                return;
+            }
+            logger.info("Performing one-time settings migration of show images and invite avatars to account data");
+            const handler = LEVEL_HANDLERS[SettingLevel.ACCOUNT];
+            const showImages = handler.getValue("showImages", null);
+            const showAvatarsOnInvites = handler.getValue("showAvatarsOnInvites", null);
+
+            if (typeof showImages === "boolean" || typeof showAvatarsOnInvites === "boolean") {
+                this.setValue("mediaPreviewConfig", null, SettingLevel.ACCOUNT, {
+                    invite_avatars: showAvatarsOnInvites === false ? MediaPreviewValue.Off : MediaPreviewValue.On,
+                    media_previews: showImages === false ? MediaPreviewValue.Off : MediaPreviewValue.On,
+                });
+            } // else, we don't set anything and use the server value
+        };
 
-        localStorage.setItem(MIGRATION_DONE_FLAG, "true");
+        doMigration().catch((e) => {
+            logger.error("Failed to migrate media config settings", e);
+        });
```

Comment:
Why do we need this nested method vs just making `migrateMediaControlsToSetting` async and inlining `doMigration`?

Reply:
I'm following what `migrateURLPreviewsE2EE` is doing, which is weirdly async yet doesn't await. I can change this however.

Reply:
Because that wrongly handles initial sync already having been completed. Copying like that is the cause of the bug you're fixing in the first place 🙃

Reply:
Yes :), I suppose I really aught to go fix that one too then